File: src/tools/voxedit/modules/voxedit-util/modifier/brush/select/Rectangle.cpp
/**
 * @file
 */

#include "Rectangle.h"
#include "Circle.h"
#include "math/Axis.h"
#include "voxedit-util/SceneManager.h"
#include "voxedit-util/modifier/ModifierVolumeWrapper.h"
#include "voxedit-util/modifier/brush/Brush.h"
#include "voxedit-util/modifier/brush/BrushGizmo.h"
#include "voxel/Face.h"
#include "voxelutil/VolumeSelect.h"
#include <glm/mat4x4.hpp>
#include <glm/trigonometric.hpp>
#include <glm/vec4.hpp>

namespace voxedit {
namespace select {

void Rectangle::computeCorners(const voxel::Region &region, voxel::FaceNames face, glm::ivec3 corners[4]) const {
	int uAxis;
	int vAxis;
	Circle::ellipseAxes(face, uAxis, vAxis);
	const int wAxis = math::getIndexForAxis(voxel::faceToAxis(face));
	const bool positiveNormal = voxel::isPositiveFace(face);
	const glm::ivec3 mins = region.getLowerCorner();
	const glm::ivec3 maxs = region.getUpperCorner();
	const float cu = (float)(mins[uAxis] + maxs[uAxis]) * 0.5f;
	const float cv = (float)(mins[vAxis] + maxs[vAxis]) * 0.5f;
	const float hu = (float)(maxs[uAxis] - mins[uAxis]) * 0.5f;
	const float hv = (float)(maxs[vAxis] - mins[vAxis]) * 0.5f;
	// Anchor the outline at the clicked (near) face depth.
	const int w0 = positiveNormal ? maxs[wAxis] : mins[wAxis];
	const float rad = glm::radians((float)_angle);
	const float cs = glm::cos(rad);
	const float sn = glm::sin(rad);
	const float local[4][2] = {{hu, hv}, {-hu, hv}, {-hu, -hv}, {hu, -hv}};
	for (int i = 0; i < 4; ++i) {
		const float lu = local[i][0];
		const float lv = local[i][1];
		glm::ivec3 p(0);
		p[uAxis] = (int)glm::round(cu + lu * cs - lv * sn);
		p[vAxis] = (int)glm::round(cv + lu * sn + lv * cs);
		p[wAxis] = w0;
		corners[i] = p;
	}
}

void Rectangle::generate(scenegraph::SceneGraph &sceneGraph, ModifierVolumeWrapper &wrapper, const BrushContext &ctx,
						 const voxel::Region &region, const AABBBrushState &state) {
	if (state.aabbFace == voxel::FaceNames::Max) {
		return;
	}
	int uAxis;
	int vAxis;
	Circle::ellipseAxes(state.aabbFace, uAxis, vAxis);
	const int wAxis = math::getIndexForAxis(voxel::faceToAxis(state.aabbFace));
	const bool positiveNormal = voxel::isPositiveFace(state.aabbFace);

	glm::ivec3 corners[4];
	computeCorners(region, state.aabbFace, corners);

	// Snap each corner to the actual surface depth at its (u,v) so the edge chord follows the
	// surface across terraces instead of staying pinned to the single clicked depth (which left
	// the down-slope edges broken at every riser).
	const voxel::Region &vol = ctx.targetVolumeRegion;
	const int w0 = corners[0][wAxis];
	const int columnTol = vol.getUpperCorner()[wAxis] - vol.getLowerCorner()[wAxis];
	for (int i = 0; i < 4; ++i) {
		int sw;
		if (voxelutil::findSurfaceNear(wrapper, corners[i][uAxis], corners[i][vAxis], w0, columnTol, uAxis, vAxis, wAxis,
									   positiveNormal, vol, sw)) {
			corners[i][wAxis] = sw;
		}
	}

	auto func = [&wrapper](int x, int y, int z, const voxel::Voxel &) {
		if (wrapper.modifierType() == ModifierType::Erase) {
			wrapper.removeFlagAt(x, y, z, voxel::FlagOutline);
		} else {
			wrapper.setFlagAt(x, y, z, voxel::FlagOutline);
		}
	};
	// Draw the four edges of the (rotated) rectangle as lines draped over the visible surface, so
	// each edge stays continuous AND visible across terraces/risers (riser faces are filled in).
	for (int i = 0; i < 4; ++i) {
		voxelutil::lineDrapeSurface(wrapper, corners[i], corners[(i + 1) % 4], uAxis, vAxis, wAxis, positiveNormal,
									_edgeWidth, vol, func);
	}
}

void Rectangle::reset() {
	Super::reset();
	_previewActive = false;
	_overlayPoints.clear();
}

void Rectangle::abort(BrushContext &ctx) {
	Super::abort(ctx);
	_previewActive = false;
	_overlayPoints.clear();
}

void Rectangle::endBrush(BrushContext &ctx) {
	Super::endBrush(ctx);
	_previewActive = false;
	_overlayPoints.clear();
}

void Rectangle::setPreview(const voxel::Region &region, voxel::FaceNames face) {
	if (face == voxel::FaceNames::Max || !region.isValid()) {
		_previewActive = false;
		_overlayPoints.clear();
		return;
	}
	glm::ivec3 corners[4];
	computeCorners(region, face, corners);
	const glm::mat4 model = _sceneManager != nullptr ? _sceneManager->worldMatrix() : glm::mat4(1.0f);
	_overlayPoints.clear();
	_overlayPoints.reserve(4);
	for (int i = 0; i < 4; ++i) {
		// Center on the voxel so the outline runs through voxel centers.
		_overlayPoints.push_back(glm::vec3(model * glm::vec4(glm::vec3(corners[i]) + 0.5f, 1.0f)));
	}
	_previewActive = true;
}

bool Rectangle::wantBrushGizmo(const BrushContext &ctx) const {
	return _previewActive && _overlayPoints.size() == 4;
}

void Rectangle::brushGizmoState(const BrushContext &ctx, BrushGizmoState &state) const {
	state.operations = BrushGizmo_WorldPolyline;
	state.worldPolyline = &_overlayPoints;
	state.worldPolylineClosed = true;
}

} // namespace select
} // namespace voxedit

File: src/tools/voxedit/modules/voxedit-util/lua/selectionmodes/rectangle.lua
function arguments()
	return {
		{name = "edgeWidth", desc = "Thickness of the rectangle outline in voxels", type = "int", default = "1", min = "1", max = "32"},
		{name = "angle", desc = "Rotation of the rectangle in the surface plane (degrees)", type = "int", default = "0", min = "-180", max = "180"},
		{
			name = "surfaceTolerance",
			desc = "How far corners may deviate from the reference depth to snap to the surface",
			type = "int",
			default = "64",
			min = "0",
			max = "64"
		}
	}
end

function description()
	return "Select a rotated rectangle outline draped over the visible surface"
end

function icon()
	return "squaredashed"
end

local sel = require("modules.selection")

local function round(n)
	if n >= 0 then
		return math.floor(n + 0.5)
	end
	return math.ceil(n - 0.5)
end

local function faceAxes(face)
	local wAxis = sel.faceToAxisIndex(face)
	local uAxis = (wAxis + 1) % 3
	local vAxis = (wAxis + 2) % 3
	return uAxis, vAxis, wAxis, sel.isPositiveFace(face)
end

local function dragRegion()
	local a = g_selectioncontext.aabbFirstPos()
	local b = g_selectioncontext.cursorPos()
	return {
		mins = function()
			return {
				x = math.min(a.x, b.x),
				y = math.min(a.y, b.y),
				z = math.min(a.z, b.z)
			}
		end,
		maxs = function()
			return {
				x = math.max(a.x, b.x),
				y = math.max(a.y, b.y),
				z = math.max(a.z, b.z)
			}
		end
	}
end

local function isZeroDrag()
	local a = g_selectioncontext.aabbFirstPos()
	local b = g_selectioncontext.cursorPos()
	return a.x == b.x and a.y == b.y and a.z == b.z
end

local function isZeroAreaOnFace(box, face)
	local uAxis, vAxis = faceAxes(face)
	local rmin = box:mins()
	local rmax = box:maxs()
	local du = sel.getAxis(rmax, uAxis) - sel.getAxis(rmin, uAxis)
	local dv = sel.getAxis(rmax, vAxis) - sel.getAxis(rmin, vAxis)
	return du == 0 and dv == 0
end

local function computeCorners(box, face, angle)
	local uAxis, vAxis, wAxis, positiveNormal = faceAxes(face)
	local rmin = box:mins()
	local rmax = box:maxs()
	local cu = (sel.getAxis(rmin, uAxis) + sel.getAxis(rmax, uAxis)) * 0.5
	local cv = (sel.getAxis(rmin, vAxis) + sel.getAxis(rmax, vAxis)) * 0.5
	local hu = (sel.getAxis(rmax, uAxis) - sel.getAxis(rmin, uAxis)) * 0.5
	local hv = (sel.getAxis(rmax, vAxis) - sel.getAxis(rmin, vAxis)) * 0.5
	local w0 = positiveNormal and sel.getAxis(rmax, wAxis) or sel.getAxis(rmin, wAxis)
	local rad = math.rad(angle)
	local cs = math.cos(rad)
	local sn = math.sin(rad)
	local localCorners = {{hu, hv}, {-hu, hv}, {-hu, -hv}, {hu, -hv}}
	local corners = {}
	for i = 1, 4 do
		local lu = localCorners[i][1]
		local lv = localCorners[i][2]
		local p = {x = 0, y = 0, z = 0}
		sel.setAxis(p, uAxis, round(cu + lu * cs - lv * sn))
		sel.setAxis(p, vAxis, round(cv + lu * sn + lv * cs))
		sel.setAxis(p, wAxis, w0)
		corners[i] = p
	end
	return corners, uAxis, vAxis, wAxis, w0
end

local function snapCornersToSurface(volume, corners, uAxis, vAxis, wAxis, w0, columnTol, face)
	for i = 1, 4 do
		local c = corners[i]
		local sw = volume:findSurfaceNear(sel.getAxis(c, uAxis), sel.getAxis(c, vAxis), w0, columnTol, face)
		if sw ~= nil then
			sel.setAxis(c, wAxis, sw)
		end
	end
end

local function outlineRectangle(volume, corners, face, edgeWidth, selecting)
	local function mark(x, y, z)
		volume:setSelected(x, y, z, selecting)
	end
	for i = 1, 4 do
		local a = corners[i]
		local b = corners[(i % 4) + 1]
		volume:lineDrapeSurface(a.x, a.y, a.z, b.x, b.y, b.z, face, edgeWidth, mark)
	end
end

local function rectangleCorners(edgeWidth, angle, surfaceTolerance, volume)
	local face = g_selectioncontext.aabbFace()
	if face == "max" or isZeroDrag() then
		return nil
	end
	local box = dragRegion()
	if isZeroAreaOnFace(box, face) then
		return nil
	end
	local corners, uAxis, vAxis, wAxis, w0 = computeCorners(box, face, angle)
	if volume ~= nil then
		local volRegion = g_selectioncontext.targetVolumeRegion()
		local rmin = volRegion:mins()
		local rmax = volRegion:maxs()
		local columnTol = sel.getAxis(rmax, wAxis) - sel.getAxis(rmin, wAxis)
		snapCornersToSurface(volume, corners, uAxis, vAxis, wAxis, w0, columnTol, face)
	end
	return corners, face
end

function gizmo(edgeWidth, angle, surfaceTolerance)
	local corners, _ = rectangleCorners(edgeWidth, angle, surfaceTolerance, nil)
	if corners == nil then
		return nil
	end
	local positions = {}
	for i = 1, 4 do
		local c = corners[i]
		positions[i] = {c.x + 0.5, c.y + 0.5, c.z + 0.5}
	end
	positions[5] = positions[1]
	return {
		operations = {"line"},
		positions = positions
	}
end

function select(node, region, edgeWidth, angle, surfaceTolerance)
	local volume = node:volume()
	local modType = g_selectioncontext.modifierType()
	local selecting = modType ~= "erase"

	local corners, face = rectangleCorners(edgeWidth, angle, surfaceTolerance, volume)
	if corners == nil then
		return
	end
	outlineRectangle(volume, corners, face, edgeWidth, selecting)
end
